index.js

```diff
@@ -282,7 +282,17 @@ async function loadPlugin ({ file, type, directoryPrefix, options, log }) {
 
   const plugin = wrapRoutes(content.default || content)
   const pluginConfig = (content.default && content.default.autoConfig) || content.autoConfig || {}
-  const pluginOptions = typeof pluginConfig === 'function' ? pluginConfig : Object.assign({}, pluginConfig, overrideConfig)
+  let pluginOptions
+  if (typeof pluginConfig === 'function') {
+    pluginOptions = function (fastify) {
+      return { ...pluginConfig(fastify), ...overrideConfig }
+    }
+
+    pluginOptions.prefix = pluginConfig.prefix
+  } else {
+    pluginOptions = { ...pluginConfig, ...overrideConfig }
+  }
+
```

Comment:
That might do the job, but I think the whole component would benefit greatly from a major refactoring.
Right now, it's really looks spaghetti code imho, it's hard to introduce changes cleanly.

Reply:
Open to work on it btw, but after the demo.

Reply:
That'd be fantastic, thx 'b

Reply:
See #383